src/main/scala/edu/berkeley/ce/rockslicing/RockSlicer.scala

```diff
@@ -129,75 +148,79 @@ object RockSlicer {
+    * along matching processor joints, eliminating false blocks caused by division of
     * input rock volume into equal volumes by load balancer
     * @param processorBlocks Sequence of blocks containing one or more processor joints each
     * @param mergedBlocks Sequence of blocks that have had processor joints removed. Initial
-    *                     input is empty sequence
-    * @param origin Origin that all blocks and their faces will be referenced to for consistency
-    * @return Sequence of blocks with all processor joints removed by merging input blocks along
-    *         matching processor joints. Redundant faces and duplicate blocks are already removed
-    *         in this list, so it is not necessary to perform these computations on the returned
-    *         sequence.
+    *                     input is an empty sequence
+    * @param orphanBlocks Sequence of blocks that cannot be matched with other blocks. Initial
+    *                     input is an empty sequence
+    * @return Tuple containing two sequences of blocks. The first sequence is all the merged blocks.
+    *         The second is all blocks that could not be merged.
     */
   @tailrec
   def mergeBlocks(processorBlocks: Seq[Block], mergedBlocks: Seq[Block],
-                  origin: (Double, Double, Double)): Seq[Block] = {
-    val reconstructedBlocks = (processorBlocks.tail map { block =>
-      val currentBlock = Block(origin, processorBlocks.head.updateFaces(origin))
-      val updatedBlock = Block(origin, block.updateFaces(origin))
-      val sharedProcFaces = compareProcessorBlocks(currentBlock, updatedBlock)
-      if (sharedProcFaces.nonEmpty) {
-        val currentFaces = currentBlock.faces.diff(sharedProcFaces)
-        val updatedFaces = updatedBlock.faces.diff(sharedProcFaces)
-        val allFaces = currentFaces ++ updatedFaces
-        // Check for any actual shared faces between blocks - if these exists blocks
-        // should not be merged since there is a real joint seperating the blocks
-        val nonSharedFaces =
-          allFaces.foldLeft(Seq.empty[Face]) { (unique, current) =>
-            if (!unique.exists(current.isSharedWith(_))) {
-              current +: unique
-            } else {
-              unique
-            }
-          }
-        if (allFaces.diff(nonSharedFaces).isEmpty) {
-          Some(Block(origin, nonSharedFaces))
-        } else {
-          None
-        }
+                  orphanBlocks: Seq[Block]):
+                 (Seq[Block], Seq[Block]) = {
+    val blockMatches = findMates(processorBlocks)
+    val joinedBlocks = blockMatches.map{ case (paired, _) => paired }
+    val matchIndices = blockMatches.flatMap { case (_, indices) => indices}.distinct
+    val originalPairedBlocks = matchIndices.collect(processorBlocks)
+
+    val uniqueBlocks = (processorBlocks ++ orphanBlocks).foldLeft(Seq.empty[Block]) { (unique, current) =>
```

Comment:
Yeah, let's definitely keep this in mind. Once we figure out the boxing issue, we can take a look using `set`. How would we implement this in `set`? Would it be the same as just using `distinct` on a `Seq`? If that's the case, we would need to write our own `Equals`...